src/types/acls.rs

```diff
@@ -0,0 +1,372 @@
+        v.sort_unstable();
+
+        write!(f, "{}", v.into_iter().collect::<String>())
+    }
+}
+
+impl From<HashSet<AclRight>> for AclRights {
+    fn from(hash: HashSet<AclRight>) -> Self {
+        Self { data: hash }
+    }
+}
+
+impl From<Vec<AclRight>> for AclRights {
+    fn from(vec: Vec<AclRight>) -> Self {
+        AclRights {
+            data: vec.into_iter().collect(),
+        }
+    }
+}
+
+impl TryFrom<&str> for AclRights {
+    type Error = AclRightError;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+        if !input
+            .chars()
+            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit())
+        {
+            return Err(AclRightError::InvalidRight);
+        }
+
+        Ok(input
+            .chars()
+            .map(|c| c.into())
+            .collect::<HashSet<AclRight>>()
+            .into())
+    }
+}
+
+/// Error from parsing AclRight strings
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum AclRightError {
+    /// Returned when a non-lower-case alpha numeric is provided in the rights list string.
+    InvalidRight,
+}
+
+impl Display for AclRightError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            AclRightError::InvalidRight => {
+                write!(f, "Rights may only be lowercase alpha numeric characters")
+            }
+        }
+    }
+}
+
+impl std::error::Error for AclRightError {}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The ACL response from the [Session::get_acl] IMAP command
+#[self_referencing]
+pub struct Acl {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: InnerAcl<'this>,
+}
+
+impl Acl {
+    /// Parse the given input into a [Acl] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        AclTryBuilder {
+            data: owned,
+            acl_builder: |input| {
+                let mut acls = Vec::new();
+
+                // There should only be ONE single ACL response
+                parse_many_into(input, &mut acls, unsolicited, |response| match response {
+                    Response::Acl(a) => Ok(MapOrNot::Map(InnerAcl {
+                        mailbox: a.mailbox,
+                        acls: a
+                            .acls
+                            .into_iter()
+                            .map(|e| AclEntry {
+                                identifier: e.identifier,
+                                rights: e.rights.into(),
+                            })
+                            .collect(),
+                    })),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })?;
+
+                if acls.is_empty() {
+                    Err(Error::Parse(ParseError::Invalid(input.to_vec())))
+                } else {
+                    Ok(acls.remove(0))
```

Comment:
Should we error if there are >1?

Reply:
I've reworked this like I did in quota so it'll Err on != 1 (and adding tests)